### PR TITLE
Rearrange comments for yaml/V3 bug fix

### PR DESCRIPTION
**Pull request by phanimarupaka**
@pwittrock 

This PR is to address the issue https://github.com/go-yaml/yaml/issues/587, as upstream might take longer as there are many tightly coupled unit tests. We can revert this once upstream is fixed.

### CODE REVIEW

File: kyaml/kio/byteio_writer.go [hunk 1]
```diff
@@ -45,6 +45,9 @@ type ByteWriter struct {
 var _ Writer = ByteWriter{}
 
 func (w ByteWriter) Write(nodes []*yaml.RNode) error {
+	for _, node := range nodes {
```
Comment (pwittrock) [inline kyaml/kio/byteio_writer.go]: Revert these changes after writing the resources

File: kyaml/yaml/types.go [hunk 2]
```diff
@@ -517,6 +518,36 @@ func String(node *yaml.Node, opts ...string) (string, error) {
 	return val, errors.Wrap(err)
 }
 
+// RearrangeComments addresses a bug in yaml V3 upstream, it parses the yaml node,
+// and rearranges the head comments of the children of sequence node.
+// Refer to https://github.com/go-yaml/yaml/issues/587 for more details
+func RearrangeComments(node *yaml.Node) {
```
Comment (pwittrock) [inline kyaml/yaml/types.go]: `ApplyHacksForSerialization(nodes []*yaml.Node) error`

File: kyaml/yaml/types.go [hunk 3]
```diff
@@ -517,6 +518,36 @@ func String(node *yaml.Node, opts ...string) (string, error) {
 	return val, errors.Wrap(err)
 }
 
+// RearrangeComments addresses a bug in yaml V3 upstream, it parses the yaml node,
```
Comment (pwittrock) [inline kyaml/yaml/types.go]: Consider putting this in its own file. -- `serialization.go`

File: kyaml/yaml/types.go [hunk 4]
```diff
@@ -517,6 +518,36 @@ func String(node *yaml.Node, opts ...string) (string, error) {
...
+
+	case MappingNode:
+		for _, node := range node.Content {
+			RearrangeComments(node)
+		}
+
+	case SequenceNode:
+		for _, node := range node.Content {
+			// for each child mapping node, transfer the head comment of it's
+			// first child scalar node to the head comment of itself
```
Comment (pwittrock) [inline kyaml/yaml/types.go]: Also

```go
// This is necessary to address serialization issue LINK_TO_ISSUE in go-yaml.v3
// Remove this hack when the issue has been resolved
```

File: kyaml/yaml/types.go [hunk 5]
```diff
@@ -517,6 +518,36 @@ func String(node *yaml.Node, opts ...string) (string, error) {
...
+			if len(node.Content) > 0 && node.Content[0].Kind == ScalarNode {
+				node.HeadComment = node.Content[0].HeadComment
+				node.Content[0].HeadComment = ""
+			}
+		}
+
+	default:
```
Comment (pwittrock) [inline kyaml/yaml/types.go]: is this necessary?

File: kyaml/kio/byteio_writer.go (hunk 1 above)
Comment (phanimarupaka) [reply kyaml/kio/byteio_writer.go]: Done

File: kyaml/yaml/types.go (hunk 3 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go]: Done

File: kyaml/yaml/types.go (hunk 4 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go]: Done

File: kyaml/yaml/types.go (hunk 5 above)
Comment (phanimarupaka) [reply kyaml/yaml/types.go]: Not necessary. Removed
